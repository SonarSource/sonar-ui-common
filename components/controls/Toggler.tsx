/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import DocumentClickHandler from './DocumentClickHandler';
import EscKeydownHandler from './EscKeydownHandler';
import OutsideClickHandler from './OutsideClickHandler';

interface Props {
  children?: React.ReactNode;
  closeOnClick?: boolean;
  closeOnClickOutside?: boolean;
  closeOnEscape?: boolean;
  onRequestClose: () => void;
  open: boolean;
  overlay: React.ReactNode;
}

export default class Toggler extends React.Component<Props> {
  renderOverlay() {
    const {
      closeOnClick = false,
      closeOnClickOutside = true,
      closeOnEscape = true,
      onRequestClose,
      overlay,
    } = this.props;

    let renderedOverlay;
    if (closeOnEscape) {
      renderedOverlay = <EscKeydownHandler onKeydown={onRequestClose}>{overlay}</EscKeydownHandler>;
    } else {
      renderedOverlay = overlay;
    }

    if (closeOnClick) {
      return (
        <DocumentClickHandler onClick={onRequestClose}>{renderedOverlay}</DocumentClickHandler>
      );
    } else if (closeOnClickOutside) {
      return (
        <OutsideClickHandler onClickOutside={onRequestClose}>{renderedOverlay}</OutsideClickHandler>
      );
    } else {
      return renderedOverlay;
    }
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.open && this.renderOverlay()}
      </>
    );
  }
}
