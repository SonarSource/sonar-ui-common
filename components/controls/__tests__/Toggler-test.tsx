/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import Toggler from '../Toggler';

it('should render only children', () => {
  expect(shallowRender({ open: false })).toMatchSnapshot();
});

it('should render children and overlay', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should render when closeOnClick=true', () => {
  expect(shallowRender({ closeOnClick: true })).toMatchSnapshot();
});

it('should not render click wrappers', () => {
  expect(
    shallowRender({ closeOnClick: false, closeOnClickOutside: false, closeOnEscape: false })
  ).toMatchSnapshot();
});

function shallowRender(props?: Partial<Toggler['props']>) {
  return shallow(
    <Toggler onRequestClose={jest.fn()} open={true} overlay={<div id="overlay" />} {...props}>
      <div id="toggle" />
    </Toggler>
  );
}
