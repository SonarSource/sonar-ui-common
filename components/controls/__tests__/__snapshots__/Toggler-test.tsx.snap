// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render click wrappers 1`] = `
<Fragment>
  <div
    id="toggle"
  />
  <div
    id="overlay"
  />
</Fragment>
`;

exports[`should render children and overlay 1`] = `
<Fragment>
  <div
    id="toggle"
  />
  <OutsideClickHandler
    onClickOutside={[MockFunction]}
  >
    <EscKeydownHandler
      onKeydown={[MockFunction]}
    >
      <div
        id="overlay"
      />
    </EscKeydownHandler>
  </OutsideClickHandler>
</Fragment>
`;

exports[`should render only children 1`] = `
<Fragment>
  <div
    id="toggle"
  />
</Fragment>
`;

exports[`should render when closeOnClick=true 1`] = `
<Fragment>
  <div
    id="toggle"
  />
  <DocumentClickHandler
    onClick={[MockFunction]}
  >
    <EscKeydownHandler
      onKeydown={[MockFunction]}
    >
      <div
        id="overlay"
      />
    </EscKeydownHandler>
  </DocumentClickHandler>
</Fragment>
`;
